### DISCUSSION TITLE
Project Building problems

### DISCUSSION THREAD

**Discussion by 2catycm** [gsantner/markor#1650] [General] [answered]
Hi, there. We are software engineering student develop team from Southern University of Science and Technology. I sincerely appreciate it that you have developed such a high-quality and easy-to-use tool. Knowing that there are many `good first issue`, we can’t wait to contribute. However, when we cloned and tried to build this project, we encountered many problems. After a week of exploration, we concluded the following problem->Solution:

### Issues

```gradle
Connection timed out: connect. If you are behind an HTTP proxy, please configure the proxy settings either in IDE or Gradle.
A problem occurred configuring root project 'markor'.
> Could not resolve all artifacts for configuration ':classpath'.
   > Could not resolve com.android.tools.build:gradle:3.6.3.
     Required by:
         project :
      > Could not resolve com.android.tools.build:gradle:3.6.3.
         > Could not get resource 'https://maven.google.com/com/android/tools/build/gradle/3.6.3/gradle-3.6.3.pom'.
            > Could not HEAD 'https://maven.google.com/com/android/tools/build/gradle/3.6.3/gradle-3.6.3.pom'.
               > Connect to maven.google.com:443 [maven.google.com/142.250.190.14] failed: Connection timed out: connect

* Try:
Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Exception is:
org.gradle.api.ProjectConfigurationException: A problem occurred configuring root project 'markor'.
	at org.gradle.configuration.project.LifecycleProjectEvaluator.wrapException(LifecycleProjectEvaluator.java:80)
	... more
```

This is because, https://maven.google.com, https://dl.google.com is inaccessible to our district due to both legal cause and server-too-far-away cause. 

To solve this, we added aliyun mirror, which is much faster in China, to download dependencies when google fails.

```groovy
repositories {
        maven { url 'https://maven.google.com' }
        jcenter()
        maven { url "https://jitpack.io" }
        mavenCentral()
		
        maven {url 'https://maven.aliyun.com/repository/public/'}
        maven {url 'https://maven.aliyun.com/repository/google/'}
        maven {url 'https://maven.aliyun.com/repository/central/'}
        maven {url 'https://maven.aliyun.com/repository/gradle-plugin/'}
        mavenLocal()
    }
```

This is modified in both buildscript{} and allproject{} in build.gradle. 

Mirror is the best solution because adding proxy or using VPN are not only illegal but also does not work. 

## Path refuse issue

```gradle
Caused by: org.gradle.api.internal.plugins.PluginApplicationException: Failed to apply plugin [id 'com.android.internal.application']
Caused by: org.gradle.api.tasks.StopExecutionException: Your project path contains non-ASCII characters. This will most likely cause the build to fail on Windows. Please move your project to a different directory. See http://b.android.com/95744 for details. This warning can be disabled by adding the line 'android.overridePathCheck=true' to gradle.properties file in the project directory.
```

Gradle refuses to build the project because the parent directory of markor (where git clone was used) contains non-ascii characters.

Solution is to add `android.overridePathCheck=true` to gradle.properties. 

## Dependencies issues
![image-20220322105206621](https://user-images.githubusercontent.com/41530341/159402281-03f9d0e2-3e40-4c8a-8fe5-0386c794b209.png)
![image-20220322111234916](https://user-images.githubusercontent.com/41530341/159402287-32d59e06-3db5-4965-a405-e1f5b4d924b3.png)
![image-20220322111453991](https://user-images.githubusercontent.com/41530341/159402290-8b7e1ed9-c651-4b74-865b-d8370d25bde6.png)
![image-20220322105149940](https://user-images.githubusercontent.com/41530341/159402294-2d653f3d-7746-4516-b74e-ed0c9b446c9e.png)

Those intricate problems can be simplified to ONE——wrong version of com/github/AppIntro/AppIntro/4.2.3 .

Especially, even we use VPN, com.github.AppIntro has no version 4.2.3 retrieved on 2022/3/16, may be it is too new. 

We tried to downgrade the version to be 4.2.2, and it seems work well and thanks goodness, no longer tedious and intractable error messages. (One of my teammate tried 4.1.0 and some new errors was produced and it is shown on above pictures too, which indicates that 4.1.0 is too low a version)

## Conclusion

Those problems are reproducible under such environment:

| Environment      | Value                                              |
| ---------------- | -------------------------------------------------- |
| Network District | Probably the whole mainland China.                 |
| Operating system | We tried Windows11,10 and Ubuntu20, no difference. |
| IDE              | We tried Android Studio and IntelliJ IDEA .        |

I think there may be two new problems if I make a PR for this:

- If aliyun mirror were added into build.gradle, will you find aliyun slow in your district just like me experiencing dramatical slow speed of google in my district? 
  - I think if we add the mirror declaration after google declaration, it would invoke mirror only when the origin does not work. 
  - Also, I haven’t find a way to let mirror settings to be a programmer-local-setting other than a project build.gradle setting. I know that maven has  settings.xml  to set aliyun, but I can’t find a similar way for gradle. 
- Do markor depends tightly on 4.2.3? Will future markor version use new features in 4.2.3 that 4.2.2 does not have? 

Looking forward to your opinions and guidance.

**Comment by gsantner** [comment, marked as answer]
Hello, welcome, great to get to know new faces in the project :handshake: 

## 1) Download issue & censorship
About download issue of standard android tooling, **censorship** & your countries restrictions I cannot help you. Raise your concerns, find a workaround or leave to other country. 

> If aliyun mirror were added into build.gradle

You and everyone who needs this is free to do so `locally`, you can have that without commiting and pushing somewhere.  
I won't merge any code/PR for loading files from some dubious server into the project. Hope this is understandable to you.  
I care about my users, their privacy and safety.

## 2) Path refuse issue
> Your project path contains non-ASCII characters

Generally in development (& also using software) you are supposed to use latin/ascii characters for filenames and paths. Many tools don't work properly without that. If you say, hey you can easy fix this, by just adding one additional flag that improves the situation with special paths ... this is fine to me if it helps.

> Solution is to add android.overridePathCheck=true to gradle.properties.

So please feel free to send a PR for that.

## 3) Dependencies issues AppIntro
Please search issues before posting, it was asked multiple times already <https://github.com/gsantner/markor/issues?q=> . It's already fixed on master branch (via #1632). So you should only need to pull from master branch and issue should be gone. 

## 4) Generally
> Hi, there. We are software engineering student develop team from Southern University of Science and Technology. I sincerely appreciate it that you have developed such a high-quality and easy-to-use tool. Knowing that there are many good first issue, we can’t wait to contribute

Thanks! Generally, if you need some help or understanding of projects/issues just ask. Warm welcome if you want to help on the project, on whatever area. Much appreciated.

**Comment by 2catycm** [comment]
Thanks for replying.&nbsp;
I can understand your concern. We will use mirror locally by our belief on it and do not commit.
Thanks for understanding, one of my friend does have special path need. If it does not work well locally, he'd better change a ascii path.
Ok, I will pull it.&nbsp;

---Original---
From: "Gregor ***@***.***&gt;
Date: Tue, Mar 22, 2022 22:45 PM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [gsantner/markor] Project Building problems (Discussion #1650)




  
I sincerely appreciate it that you have developed such a high-quality and easy-to-use tool.
  
Hello, welcome, great to get to know new faces in the project 🤝
 
1) Download issue &amp; censorship
 
About download issue of standard android tooling, censorship &amp; your countries restrictions I cannot help you. Raise your concerns, find a workaround or leave to other country.
  
If aliyun mirror were added into build.gradle
  
You and everyone who needs this is free to do so locally, you can have that without commiting and pushing somewhere.
 I won't merge any code/PR for loading files from some dubious server into the project. Hope this is understandable to you.
 I care about my users, their privacy and safety.
 
2) Path refuse issue
  
Your project path contains non-ASCII characters
  
Generally in development (&amp; also using software) you are supposed to use latin/ascii characters for filenames and paths. Many tools don't work properly without that. If you say, hey you can easy fix this, by just adding one additional flag that improves the situation with special paths ... this is fine to me if it helps.
  
Solution is to add android.overridePathCheck=true to gradle.properties.
  
So please feel free to send a PR for that.
 
3) Dependencies issues AppIntro
 
Please search issues before posting, it was asked multiple times already https://github.com/gsantner/markor/issues?q= . It's already fixed on master branch (via #1632). So you should only need to pull from master branch and issue should be gone.
 
4) Generally
  
Hi, there. We are software engineering student develop team from Southern University of Science and Technology. I sincerely appreciate it that you have developed such a high-quality and easy-to-use tool. Knowing that there are many good first issue, we can’t wait to contribute
  
Generally, if you need some help or understanding of projects/issues just ask. Warm welcome if you want to help on the project, on whatever area. Much appreciated.
